### ISSUE TITLE
Enhancement: [ban-types] ban certain types to be extended from

### ISSUE THREAD

**Issue by Haberkamp** [typescript-eslint/typescript-eslint#7127]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/ban-types/

### Description

I think that it would be useful to ban certain types from being extended. The job of the rule is to ban the usage of certain types and provide an alternative way if possible. Creating a new interface and extending from a harmful/confusing type also will introduce the usage of that not wanted type.

One use case might be to eliminate the usage of `HTMLProps` and `HTMLAttributes`.

### Fail

```typescript
interface Props extends from HTMLProps<HTMLButtonElement> {}
```

### Pass

```typescript
interface Props extends from ComponentPropsWithoutRef<'div>' {}
```

### Additional Info

Link to the resource on why `HTMLProps` and `HTMLAttributes` can be problematic: https://react-typescript-cheatsheet.netlify.app/docs/advanced/patterns_by_usecase/#wrappingmirroring-a-html-element

**Comment by bradzacher** [comment]
I definitely understand the motivation behind this - though I think that the applicability of such a feature is very niche.
There are very few cases where you absolutely never want to use a type just within an `extends` but are okay with it everywhere else. 

As an additional note you would also need to handle intersections (which are essentially the same as interface extends).

If we had some more usecases for this feature I think we could consider it, but with a single narrow case for one particular framework's set of types - not enough.

If you're just looking to target the extends case this can be easily handled using `no-restricted-syntax` - https://typescript-eslint.io/linting/troubleshooting#how-can-i-ban-specific-language-feature

**Comment by Haberkamp** [comment]
I've come up with another case. What if an interface, the team has created, on a large codebase, has been deprecated and an alternative way is provided. Now in order to not introduce the old interface in new places of the codebase, they decide to ban the type, add comments to the existing places where the interface is used to suppress errors and incrementally, over time replace the old interface with the new one. Interested to hear your thoughts on that. 

Also if I'm not mistaken intersections are already handled by this rule. When I write the following code eslint will throw an error:

```typescript
type MyType = {
  value: boolean;
} & HTMLProps<HTMLButtonElement>;
```
The last line will be marked with an error, telling the that I cannot use the HTMLProps type.

**Comment by bradzacher** [comment]
Okay, sorry, I misunderstood the intent of your issue.

I thought you to be asking for a feature where you would ban a type specifically in an `extends` location, but nowhere else.

However instead you meant you would like the rule to cover extends, [when it currently does not](https://typescript-eslint.io/play/#ts=5.0.4&jsx=true&sourceType=module&code=JYOwLgpgTgZghgYwgAgGIHt3IN7IL4BQBoksiKAQnFMhAB6QgAmAzmpjvgWAJ4AOlOAC9kAXmQBGZADJ26ANwEEAGzgs2VALbJgmvsoiaI4Nhiy5CQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AIwEMnajTUMAbXDYciaNAD20SABoRorKOyQ+KdJgXKcAMUmSN+OIi2iAvqbAXRAXREWzQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

The former is a no-go, but the latter makes sense.
I don't remember why the rule behaves this way.. I am 95% sure it's an oversight that nobody has ever noticed.
I don't see a reason why it shouldn't cover these cases!

**Comment by Haberkamp** [comment]
Awesome, I'll try to open a PR for that! 👍
